Metalama.Framework.Engine/Utilities/DeclarationHelper.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) SharpCrafters s.r.o. See the LICENSE.md file in the root directory of this repository root for details.
+
+using Metalama.Framework.Code;
+using Metalama.Framework.Code.Types;
+using Metalama.Framework.Engine.CodeModel;
+using Metalama.Framework.Engine.Utilities.Caching;
+using Microsoft.CodeAnalysis;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace Metalama.Framework.Engine.Utilities;
+
+internal static class DeclarationHelper
+{
+    private enum TypeNameKind
+    {
+        Name,
+        FullName
+    }
+
+    private static readonly WeakCache<IType, string> _reflectionNameCache = new();
+    private static readonly WeakCache<IType, string> _reflectionFullNameCache = new();
+
+    /// <summary>
+    /// Gets a string that would be equal to <see cref="MemberInfo.Name"/>.
+    /// </summary>
+    internal static string GetReflectionName( this IType type )
+        => _reflectionNameCache.GetOrAdd( type, x => x.GetReflectionName( TypeNameKind.Name ) );
+
+    /// <summary>
+    /// Gets a string that would be equal to <see cref="Type.FullName"/>, except that we do not qualify type names with the assembly name.
+    /// </summary>
+    internal static string GetReflectionFullName( this IType type )
+        => _reflectionFullNameCache.GetOrAdd( type, x => x.GetReflectionName( TypeNameKind.FullName ) );
+
+    private static string GetReflectionName( this IType declaration, TypeNameKind kind )
+    {
+        if ( declaration is ITypeParameter typeParameter )
```

Comment:
The implementation should fall back to an ISymbol-based one if one is available to avoid allocating the IDeclaration of parents.